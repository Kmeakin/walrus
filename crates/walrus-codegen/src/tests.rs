use crate::codegen::Context;
use inkwell::{memory_buffer::MemoryBuffer, OptimizationLevel};
use insta::*;

#[track_caller]
fn test_codegen(src: &str) { test::<()>(src, None) }

#[track_caller]
fn test_codegen_and_run<T>(src: &str, expected: T)
where
    T: PartialEq + std::fmt::Debug,
{
    test(src, Some(expected))
}

fn test<T>(src: &str, expected: Option<T>)
where
    T: PartialEq + std::fmt::Debug,
{
    let syntax = walrus_parser::parse(src);
    let hir = walrus_semantics::hir::lower(&syntax);
    let scopes = walrus_semantics::scopes::scopes(&hir);
    let types = walrus_semantics::ty::infer(hir.clone(), scopes.clone());

    dbg!(&syntax);
    dbg!(&hir);
    dbg!(&scopes);
    dbg!(&types);

    let module = crate::codegen("module", hir.hir, scopes, types);

    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("../snapshots");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| assert_display_snapshot!(module));

    let llvm = Context::create();
    let module_buffer = MemoryBuffer::create_from_memory_range_copy(module.as_bytes(), "module");
    let llvm_module = llvm.create_module_from_ir(module_buffer).unwrap();

    if let Some(expected) = expected {
        let exec_engine = llvm_module
            .create_jit_execution_engine(OptimizationLevel::None)
            .unwrap();
        let f = unsafe { exec_engine.get_function::<unsafe extern "C" fn() -> T>("main") }.unwrap();
        assert_eq!(unsafe { f.call() }, expected);
    }
}

mod literals {
    use super::*;
    #[test]
    fn empty_fn() { test_codegen_and_run("fn main() -> _ {}", ()) }

    #[test]
    fn lit_true() { test_codegen_and_run(r#"fn main() -> _ { true }"#, true) }

    #[test]
    fn lit_false() { test_codegen_and_run(r#"fn main() -> _ { false }"#, false) }

    #[test]
    fn lit_int() { test_codegen_and_run(r#"fn main() -> _ { 1 }"#, 1_i32) }

    #[test]
    fn lit_float() { test_codegen_and_run(r#"fn main() -> _ { 1.234 }"#, 1.234_f32) }

    #[test]
    fn lit_char() { test_codegen_and_run(r#"fn main() -> _ { 'a' }"#, 'a') }

    #[test]
    fn lit_string() {
        test_codegen_and_run(r#"fn main() -> _ { "hello world" }"#, "hello world".len())
    }
}

mod tuples {
    use super::*;
    #[test]
    fn tuple0() { test_codegen_and_run(r#"fn main() -> _ { () }"#, ()) }

    #[test]
    fn tuple1() { test_codegen_and_run(r#"fn main() -> _ { (1,) }"#, (1_i32,)) }

    #[test]
    fn tuple2() { test_codegen_and_run(r#"fn main() -> _ { (1, 2) }"#, (1_i32, 2_i64)) }
}

mod if_ {
    use super::*;
    #[test]
    fn if_then_else_true() {
        test_codegen_and_run("fn main() -> _ { if true {1} else {0} }", 1_i32)
    }

    #[test]
    fn if_then_else_false() {
        test_codegen_and_run("fn main() -> _ { if false {1} else {0} }", 0_i32)
    }

    #[test]
    fn if_then_true() { test_codegen_and_run(r#"fn main() -> _ { if true {1} }"#, ()); }
    #[test]
    fn if_then_false() { test_codegen_and_run(r#"fn main() -> _ { if false {1} }"#, ()); }
}

mod let_ {
    use super::*;

    #[test]
    fn let_var() { test_codegen_and_run(r#"fn main() -> _ { let x = 5; x }"#, 5_i32); }

    #[test]
    fn let_tuple2() { test_codegen_and_run(r#"fn main() -> _ { let (x, y) = (5, 6); x }"#, 5_i32) }

    #[test]
    fn let_tuple4() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    let (a, b, c, d) = (9, 8, 7, 6);
    let (x, (y, z), w) = (d, (c, b), a);
    x
}"#,
            6_i32,
        )
    }
}

mod fns {
    use super::*;
    #[test]
    fn id_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { id(5) }
fn id(x: _) -> _ {x}
"#,
            5_i32,
        )
    }

    #[test]
    fn get_five_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { get_five() }
fn get_five() -> _ {5}
"#,
            5_i64,
        )
    }

    #[test]
    fn fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = get_five;
    f()
}

fn get_five() -> _ {5}
"#,
            5_i32,
        )
    }

    #[test]
    fn builtin_fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = putchar;
    f('a')
}
"#,
            (),
        )
    }
}

mod return_ {
    use super::*;
    #[test]
    fn early_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    6
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multiple_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    return 6;
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multi_expr_return() { test_codegen_and_run(r#"fn main() -> _ { 1 + (return 2) }"#, 2); }

    #[test]
    #[ignore]
    fn if_then_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    if true {return 1}
    else {0}
}
"#,
            1,
        )
    }
}

mod bool_ops {
    use super::*;
    #[test]
    fn bool_not() { test_codegen_and_run(r#"fn main() -> _  {!false}"#, true) }
    #[test]
    fn bool_or() { test_codegen_and_run(r#"fn main() -> _  {false || true}"#, true) }

    #[test]
    fn bool_and() { test_codegen_and_run(r#"fn main() -> _ {true && false}"#, false) }
}

mod int_ops {
    use super::*;

    #[test]
    fn int_neg() { test_codegen_and_run(r#"fn main() ->   _  {-1}"#, -1_i32) }
    #[test]
    fn int_plus() { test_codegen_and_run(r#"fn main() ->   _  {+1}"#, 1_i32) }

    #[test]
    fn int_add() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x + y}"#, 2_i32) }
    #[test]
    fn int_sub() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x - y}"#, 0_i32) }
    #[test]
    fn int_mul() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 2; x * y}"#, 2_i32) }
    #[test]
    fn int_div() { test_codegen_and_run(r#"fn main() -> _ {let x = 5; let y = 2; x / y}"#, 2_i32) }

    #[test]
    fn int_eq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x == y}"#, true) }
    #[test]
    fn int_neq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x != y}"#, false) }
    #[test]
    fn int_less() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x < y}"#, false) }
    #[test]
    fn int_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x <= y}"#, true)
    }
    #[test]
    fn int_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x > y}"#, false)
    }
    #[test]
    fn int_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x >= y}"#, true)
    }
}

mod float_ops {
    use super::*;

    #[test]
    fn float_neg() { test_codegen_and_run(r#"fn main() -> _  {-1.0}"#, -1.0_f32) }
    #[test]
    fn float_plus() { test_codegen_and_run(r#"fn main() -> _  {+1.0}"#, 1.0_f32) }

    #[test]
    fn float_add() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x + y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_sub() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x - y}"#,
            0.0_f32,
        )
    }
    #[test]
    fn float_mul() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 2.0; x * y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_div() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 5.0; let y = 2.0; x / y}"#,
            2.5_f32,
        )
    }

    #[test]
    fn float_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x == y}"#, true)
    }
    #[test]
    fn float_neq() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x != y}"#,
            false,
        )
    }
    #[test]
    fn float_less() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x < y}"#, false)
    }
    #[test]
    fn float_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x <= y}"#, true)
    }
    #[test]
    fn float_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x > y}"#, false)
    }
    #[test]
    fn float_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x >= y}"#, true)
    }
}

mod assign {
    use super::*;

    #[test]
    fn assign_var() {
        test_codegen_and_run(
            r#"fn main() -> _ {
        let mut x = 5;
        x = 6;
        x
    }"#,
            6_i32,
        )
    }

    #[test]
    fn assign_field() {
        test_codegen_and_run(
            r#"
            struct S {x: Int}

            fn main() -> _ {
                let mut s = S {x: 0};
                s.x = 1;
                s.x
        }"#,
            1_i32,
        )
    }
}

mod builtins {
    use super::*;

    #[test]
    fn builtin_exit() { test_codegen("fn main() -> Int { exit(1) }") }

    #[test]
    fn builtin_putchar() {
        test_codegen_and_run(
            r#"
fn main() -> () {
    putchar('h');
    putchar('e');
    putchar('l');
    putchar('l');
    putchar('o');
    putchar(' ');
    putchar('w');
    putchar('o');
    putchar('r');
    putchar('l');
    putchar('d');
}
"#,
            (),
        )
    }
}

mod loops {
    use super::*;

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_forever() { test_codegen("fn main() -> Int { loop {} }") }

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_and_continue() { test_codegen("fn main() -> () { loop {let x = 5; continue} }") }

    #[test]
    fn loop_and_break() {
        test_codegen_and_run("fn main() -> _ { loop {let x = 5; break x} }", 5_i32)
    }

    #[test]
    #[ignore]
    fn loop_and_break_and_continue() {
        test_codegen_and_run(
            "fn main() -> _ { loop { if true {continue} else {break} } }",
            (),
        )
    }
}

mod structs {
    use super::*;
    #[test]
    fn struct_constructor() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int}
fn main() -> _ {
    let foo = Foo {x:5};
    foo
}
"#,
            5_i32,
        )
    }

    #[test]
    fn field_expr() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int,y:Int}
fn main() -> _ {
    let foo = Foo {y:6, x:5};
    foo.y
}
"#,
            6_i32,
        )
    }

    #[test]
    fn struct_pat() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int, y: Int}
fn main() -> _ {
    let Foo {y,x} = Foo{x: 5, y: 10};
    y - x
}
    "#,
            5_i32,
        );
    }

    #[test]
    fn struct_in_struct() {
        test_codegen_and_run(
            r#"
struct Foo {bar: Bar}
struct Bar {x: Int, y: Int}
fn main() -> _ {
    let bar = Bar{x: 5, y: 10};
    let foo = Foo{bar: bar};
    foo.bar.y - foo.bar.x
}
    "#,
            5_i32,
        );
    }
}

mod enums {
    use super::*;

    #[test]
    fn enum0() {
        test_codegen(
            r#"
        enum Void {}
        fn main() -> _ {
            let impossible: Void = exit(0);
            impossible
        }
    "#,
        )
    }

    #[test]
    fn enum1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {} 
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            ((), ()),
        )
    }

    #[test]
    fn enum1_with_payload() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (5_i32, ()),
        )
    }

    #[test]
    fn enum2_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            (0_u8, ()),
        );
    }
    #[test]
    fn enum2_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Baz{}
        }
    "#,
            (1_u8, ()),
        )
    }

    #[test]
    fn enum2_with_payload_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int},
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (0_u8, 5_i64),
        )
    }

    #[test]
    #[ignore]
    fn enum2_with_payload_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int}, 
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Baz{y: true}
        }
    "#,
            (1_u8, 1_i32),
        )
    }
}

mod pattern_matching {
    use super::*;

    #[test]
    fn simple_pattern_match() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn main() -> _ {
    let opt = Option::Some{x:1};
    match opt {
        Option::None{} => 0,
        Option::Some{x} => x,
    }
}
    "#,
            1_i32,
        )
    }

    #[test]
    fn add_options() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn add_options(a: Option, b: Option) -> Option {
    match (a, b) {
        (Option::Some{x:a}, Option::Some{x:b}) => Option::Some{x: a + b},
        _ => Option::None{},
    }
}

fn main() -> _ {
    let a = Option::Some{x:1};
    let b = Option::Some{x:2};
    add_options(a,b)
}
    "#,
            (1_i8, 3_i64),
        )
    }

    #[test]
    fn list_len() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}

fn length(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head: _, tail} => 1 + length(tail),
    }
}

fn main() -> _ {
    let l = List::Cons{head: 2, tail: List::Cons {head: 1, tail: List::Nil{}}};
    length(l)
}
        "#,
            2_i32,
        )
    }

    #[test]
    fn list_map() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn main() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }

    #[test]
    fn danglin_ptr() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn get_list() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    l1
}

fn main() -> _ {
    let l1 = get_list();
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }
}
