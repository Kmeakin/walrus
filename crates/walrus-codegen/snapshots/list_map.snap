---
source: crates/walrus-codegen/src/tests.rs
expression: module

---
; ModuleID = 'module'
source_filename = "module"

%List = type { i8, { i32, %List* } }

define %List @map(%List %map.params.0, { i32 (i8*, i32)*, i8* } %map.params.1) {
map.entry:
  %l.alloca = alloca %List, align 8
  store %List %map.params.0, %List* %l.alloca, align 8
  %f.alloca = alloca { i32 (i8*, i32)*, i8* }, align 8
  store { i32 (i8*, i32)*, i8* } %map.params.1, { i32 (i8*, i32)*, i8* }* %f.alloca, align 8
  %l = load %List, %List* %l.alloca, align 8
  br label %match.case0.test

match.case0.test:                                 ; preds = %map.entry
  %List.discriminant3 = extractvalue %List %l, 0
  %"List::Nil.cmp_discriminant" = icmp eq i8 %List.discriminant3, 0
  br i1 %"List::Nil.cmp_discriminant", label %"match.case0.List::Nil.then", label %"match.case0.List::Nil.end"

"match.case0.List::Nil.then":                     ; preds = %match.case0.test
  %List.payload4 = extractvalue %List %l, 1
  br label %"match.case0.List::Nil.end"

"match.case0.List::Nil.end":                      ; preds = %"match.case0.List::Nil.then", %match.case0.test
  %"match.case0.List::Nil.phi" = phi i1 [ true, %"match.case0.List::Nil.then" ], [ false, %match.case0.test ]
  br i1 %"match.case0.List::Nil.phi", label %match.case0.then, label %match.case1.test

match.case0.then:                                 ; preds = %"match.case0.List::Nil.end"
  %List.payload5 = extractvalue %List %l, 1
  %List.alloca6 = alloca %List, align 8
  %List.discriminant.gep7 = getelementptr inbounds %List, %List* %List.alloca6, i32 0, i32 0
  store i8 0, i8* %List.discriminant.gep7, align 1
  %List.payload.gep8 = getelementptr inbounds %List, %List* %List.alloca6, i32 0, i32 1
  %List.load9 = load %List, %List* %List.alloca6, align 8
  br label %match.end

match.case1.test:                                 ; preds = %"match.case0.List::Nil.end"
  %List.discriminant = extractvalue %List %l, 0
  %"List::Cons.cmp_discriminant" = icmp eq i8 %List.discriminant, 1
  br i1 %"List::Cons.cmp_discriminant", label %"match.case1.List::Cons.then", label %"match.case1.List::Cons.end"

"match.case1.List::Cons.then":                    ; preds = %match.case1.test
  %List.payload = extractvalue %List %l, 1
  br label %"match.case1.List::Cons.end"

"match.case1.List::Cons.end":                     ; preds = %"match.case1.List::Cons.then", %match.case1.test
  %"match.case1.List::Cons.phi" = phi i1 [ true, %"match.case1.List::Cons.then" ], [ false, %match.case1.test ]
  br i1 %"match.case1.List::Cons.phi", label %match.case1.then, label %match.fail

match.case1.then:                                 ; preds = %"match.case1.List::Cons.end"
  %List.payload1 = extractvalue %List %l, 1
  %"List::Cons.head" = extractvalue { i32, %List* } %List.payload1, 0
  %head.alloca = alloca i32, align 4
  store i32 %"List::Cons.head", i32* %head.alloca, align 4
  %"List::Cons.tail" = extractvalue { i32, %List* } %List.payload1, 1
  %"List::Cons.tail.load" = load %List, %List* %"List::Cons.tail", align 8
  %tail.alloca = alloca %List, align 8
  store %List %"List::Cons.tail.load", %List* %tail.alloca, align 8
  %List.alloca = alloca %List, align 8
  %List.discriminant.gep = getelementptr inbounds %List, %List* %List.alloca, i32 0, i32 0
  store i8 1, i8* %List.discriminant.gep, align 1
  %List.payload.gep = getelementptr inbounds %List, %List* %List.alloca, i32 0, i32 1
  %head = load i32, i32* %head.alloca, align 4
  %f = load { i32 (i8*, i32)*, i8* }, { i32 (i8*, i32)*, i8* }* %f.alloca, align 8
  %closure.code = extractvalue { i32 (i8*, i32)*, i8* } %f, 0
  %closure.env = extractvalue { i32 (i8*, i32)*, i8* } %f, 1
  %lambda.call = call i32 %closure.code(i8* %closure.env, i32 %head)
  %"List::Cons.head.gep" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep, i32 0, i32 0
  store i32 %lambda.call, i32* %"List::Cons.head.gep", align 4
  %tail = load %List, %List* %tail.alloca, align 8
  %f2 = load { i32 (i8*, i32)*, i8* }, { i32 (i8*, i32)*, i8* }* %f.alloca, align 8
  %map.call = call %List @map(%List %tail, { i32 (i8*, i32)*, i8* } %f2)
  %"List::Cons.tail.gep" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep, i32 0, i32 1
  %malloccall = tail call i8* @malloc(i32 ptrtoint (%List* getelementptr (%List, %List* null, i32 1) to i32))
  %"List::Cons.tail.malloc" = bitcast i8* %malloccall to %List*
  store %List %map.call, %List* %"List::Cons.tail.malloc", align 8
  store %List* %"List::Cons.tail.malloc", %List** %"List::Cons.tail.gep", align 8
  %List.load = load %List, %List* %List.alloca, align 8
  br label %match.end

match.fail:                                       ; preds = %"match.case1.List::Cons.end"
  unreachable

match.end:                                        ; preds = %match.case0.then, %match.case1.then
  %match.phi = phi %List [ %List.load, %match.case1.then ], [ %List.load9, %match.case0.then ]
  ret %List %match.phi
}

define i32 @sum(%List %sum.params.0) {
sum.entry:
  %l.alloca = alloca %List, align 8
  store %List %sum.params.0, %List* %l.alloca, align 8
  %l = load %List, %List* %l.alloca, align 8
  br label %match.case0.test

match.case0.test:                                 ; preds = %sum.entry
  %List.discriminant2 = extractvalue %List %l, 0
  %"List::Nil.cmp_discriminant" = icmp eq i8 %List.discriminant2, 0
  br i1 %"List::Nil.cmp_discriminant", label %"match.case0.List::Nil.then", label %"match.case0.List::Nil.end"

"match.case0.List::Nil.then":                     ; preds = %match.case0.test
  %List.payload3 = extractvalue %List %l, 1
  br label %"match.case0.List::Nil.end"

"match.case0.List::Nil.end":                      ; preds = %"match.case0.List::Nil.then", %match.case0.test
  %"match.case0.List::Nil.phi" = phi i1 [ true, %"match.case0.List::Nil.then" ], [ false, %match.case0.test ]
  br i1 %"match.case0.List::Nil.phi", label %match.case0.then, label %match.case1.test

match.case0.then:                                 ; preds = %"match.case0.List::Nil.end"
  %List.payload4 = extractvalue %List %l, 1
  br label %match.end

match.case1.test:                                 ; preds = %"match.case0.List::Nil.end"
  %List.discriminant = extractvalue %List %l, 0
  %"List::Cons.cmp_discriminant" = icmp eq i8 %List.discriminant, 1
  br i1 %"List::Cons.cmp_discriminant", label %"match.case1.List::Cons.then", label %"match.case1.List::Cons.end"

"match.case1.List::Cons.then":                    ; preds = %match.case1.test
  %List.payload = extractvalue %List %l, 1
  br label %"match.case1.List::Cons.end"

"match.case1.List::Cons.end":                     ; preds = %"match.case1.List::Cons.then", %match.case1.test
  %"match.case1.List::Cons.phi" = phi i1 [ true, %"match.case1.List::Cons.then" ], [ false, %match.case1.test ]
  br i1 %"match.case1.List::Cons.phi", label %match.case1.then, label %match.fail

match.case1.then:                                 ; preds = %"match.case1.List::Cons.end"
  %List.payload1 = extractvalue %List %l, 1
  %"List::Cons.head" = extractvalue { i32, %List* } %List.payload1, 0
  %head.alloca = alloca i32, align 4
  store i32 %"List::Cons.head", i32* %head.alloca, align 4
  %"List::Cons.tail" = extractvalue { i32, %List* } %List.payload1, 1
  %"List::Cons.tail.load" = load %List, %List* %"List::Cons.tail", align 8
  %tail.alloca = alloca %List, align 8
  store %List %"List::Cons.tail.load", %List* %tail.alloca, align 8
  %head = load i32, i32* %head.alloca, align 4
  %tail = load %List, %List* %tail.alloca, align 8
  %sum.call = call i32 @sum(%List %tail)
  %Int.add = add i32 %head, %sum.call
  br label %match.end

match.fail:                                       ; preds = %"match.case1.List::Cons.end"
  unreachable

match.end:                                        ; preds = %match.case0.then, %match.case1.then
  %match.phi = phi i32 [ %Int.add, %match.case1.then ], [ 0, %match.case0.then ]
  ret i32 %match.phi
}

define i32 @main() {
main.entry:
  %List.alloca = alloca %List, align 8
  %List.discriminant.gep = getelementptr inbounds %List, %List* %List.alloca, i32 0, i32 0
  store i8 1, i8* %List.discriminant.gep, align 1
  %List.payload.gep = getelementptr inbounds %List, %List* %List.alloca, i32 0, i32 1
  %"List::Cons.head.gep" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep, i32 0, i32 0
  store i32 4, i32* %"List::Cons.head.gep", align 4
  %List.alloca1 = alloca %List, align 8
  %List.discriminant.gep2 = getelementptr inbounds %List, %List* %List.alloca1, i32 0, i32 0
  store i8 1, i8* %List.discriminant.gep2, align 1
  %List.payload.gep3 = getelementptr inbounds %List, %List* %List.alloca1, i32 0, i32 1
  %"List::Cons.head.gep4" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep3, i32 0, i32 0
  store i32 3, i32* %"List::Cons.head.gep4", align 4
  %List.alloca5 = alloca %List, align 8
  %List.discriminant.gep6 = getelementptr inbounds %List, %List* %List.alloca5, i32 0, i32 0
  store i8 1, i8* %List.discriminant.gep6, align 1
  %List.payload.gep7 = getelementptr inbounds %List, %List* %List.alloca5, i32 0, i32 1
  %"List::Cons.head.gep8" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep7, i32 0, i32 0
  store i32 2, i32* %"List::Cons.head.gep8", align 4
  %List.alloca9 = alloca %List, align 8
  %List.discriminant.gep10 = getelementptr inbounds %List, %List* %List.alloca9, i32 0, i32 0
  store i8 1, i8* %List.discriminant.gep10, align 1
  %List.payload.gep11 = getelementptr inbounds %List, %List* %List.alloca9, i32 0, i32 1
  %"List::Cons.head.gep12" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep11, i32 0, i32 0
  store i32 1, i32* %"List::Cons.head.gep12", align 4
  %List.alloca13 = alloca %List, align 8
  %List.discriminant.gep14 = getelementptr inbounds %List, %List* %List.alloca13, i32 0, i32 0
  store i8 0, i8* %List.discriminant.gep14, align 1
  %List.payload.gep15 = getelementptr inbounds %List, %List* %List.alloca13, i32 0, i32 1
  %List.load = load %List, %List* %List.alloca13, align 8
  %"List::Cons.tail.gep" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep11, i32 0, i32 1
  %malloccall = tail call i8* @malloc(i32 ptrtoint (%List* getelementptr (%List, %List* null, i32 1) to i32))
  %"List::Cons.tail.malloc" = bitcast i8* %malloccall to %List*
  store %List %List.load, %List* %"List::Cons.tail.malloc", align 8
  store %List* %"List::Cons.tail.malloc", %List** %"List::Cons.tail.gep", align 8
  %List.load16 = load %List, %List* %List.alloca9, align 8
  %"List::Cons.tail.gep17" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep7, i32 0, i32 1
  %malloccall18 = tail call i8* @malloc(i32 ptrtoint (%List* getelementptr (%List, %List* null, i32 1) to i32))
  %"List::Cons.tail.malloc19" = bitcast i8* %malloccall18 to %List*
  store %List %List.load16, %List* %"List::Cons.tail.malloc19", align 8
  store %List* %"List::Cons.tail.malloc19", %List** %"List::Cons.tail.gep17", align 8
  %List.load20 = load %List, %List* %List.alloca5, align 8
  %"List::Cons.tail.gep21" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep3, i32 0, i32 1
  %malloccall22 = tail call i8* @malloc(i32 ptrtoint (%List* getelementptr (%List, %List* null, i32 1) to i32))
  %"List::Cons.tail.malloc23" = bitcast i8* %malloccall22 to %List*
  store %List %List.load20, %List* %"List::Cons.tail.malloc23", align 8
  store %List* %"List::Cons.tail.malloc23", %List** %"List::Cons.tail.gep21", align 8
  %List.load24 = load %List, %List* %List.alloca1, align 8
  %"List::Cons.tail.gep25" = getelementptr inbounds { i32, %List* }, { i32, %List* }* %List.payload.gep, i32 0, i32 1
  %malloccall26 = tail call i8* @malloc(i32 ptrtoint (%List* getelementptr (%List, %List* null, i32 1) to i32))
  %"List::Cons.tail.malloc27" = bitcast i8* %malloccall26 to %List*
  store %List %List.load24, %List* %"List::Cons.tail.malloc27", align 8
  store %List* %"List::Cons.tail.malloc27", %List** %"List::Cons.tail.gep25", align 8
  %List.load28 = load %List, %List* %List.alloca, align 8
  %l1.alloca = alloca %List, align 8
  store %List %List.load28, %List* %l1.alloca, align 8
  %l1 = load %List, %List* %l1.alloca, align 8
  %closure.alloca = alloca { i32 (i8*, i32)*, i8* }, align 8
  %closure.code = getelementptr inbounds { i32 (i8*, i32)*, i8* }, { i32 (i8*, i32)*, i8* }* %closure.alloca, i32 0, i32 0
  store i32 (i8*, i32)* @lambda, i32 (i8*, i32)** %closure.code, align 8
  %closure.env = getelementptr inbounds { i32 (i8*, i32)*, i8* }, { i32 (i8*, i32)*, i8* }* %closure.alloca, i32 0, i32 1
  %malloccall29 = tail call i8* @malloc(i32 0)
  %env.malloc = bitcast i8* %malloccall29 to {}*
  %env = bitcast {}* %env.malloc to i8*
  store i8* %env, i8** %closure.env, align 8
  %closure = load { i32 (i8*, i32)*, i8* }, { i32 (i8*, i32)*, i8* }* %closure.alloca, align 8
  %map.call = call %List @map(%List %l1, { i32 (i8*, i32)*, i8* } %closure)
  %l2.alloca = alloca %List, align 8
  store %List %map.call, %List* %l2.alloca, align 8
  %l2 = load %List, %List* %l2.alloca, align 8
  %sum.call = call i32 @sum(%List %l2)
  %s.alloca = alloca i32, align 4
  store i32 %sum.call, i32* %s.alloca, align 4
  %s = load i32, i32* %s.alloca, align 4
  ret i32 %s
}

declare noalias i8* @malloc(i32)

define i32 @lambda(i8* %env_ptr, i32 %params.0) {
lambda.entry:
  %env_ptr1 = bitcast i8* %env_ptr to {}*
  %env = load {}, {}* %env_ptr1, align 1
  %x.alloca = alloca i32, align 4
  store i32 %params.0, i32* %x.alloca, align 4
  %x = load i32, i32* %x.alloca, align 4
  %x2 = load i32, i32* %x.alloca, align 4
  %Int.mul = mul i32 %x, %x2
  ret i32 %Int.mul
}

